### PR TITLE
feat(middleware): add a middleware to track the state of messages. fix #106

### LINKED ISSUE #106: Create a status middleware
First thing is people want to see which message are being processed.

We should start on 5 status:
- pending
- started
- canceled
- success
- failure
- skiped

All errors should be caught, as it's not very important if we fail to monitor a message.

Store the status of each message in Redis (default result_ttl pretty long => 24h).

## Checklist

* [ ] Create the middleware class
* [ ] Update the message status on after_enqueue, before_process_message, after_process_message, after_message_canceled, ...
* [ ] store the arguments with the message (check if they are not too big in size)

### PR DESCRIPTION
**Pull request by xdanielsb**

### CODE REVIEW

File: tests/middleware/test_message_state.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+
+    """ TODO: find alternative to change priority of rabbitmq_broker in order to
+           use it in this test"""
+    def test_pending_state_message(self, rabbitmq_state, redis_state_backend):
+        rabbitmq_state.declare_actor(TestMessageState.add)
+        rabbitmq_state.add_middleware(MessageState(backend=redis_state_backend))
+        msg = TestMessageState.add.send(5, 7)
+        assert redis_state_backend._get(msg.message_id)['state'] == "Pending"
+
+    def test_success_state_message(self, local_broker, redis_state_backend):
```
Comment (antoinerabany) [inline tests/middleware/test_message_state.py:31]: do not use local_broker, this test should look like `test_failure_state_message` but with an actor which does not raise

File: tests/middleware/test_message_state.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+        stub_broker.join(TestMessageState.error.queue_name)
+        stub_worker.join()
+        assert redis_state_backend._get(msg.message_id)['state'] == "Failure"
+
+    def test_cancel_state_message(self, stub_broker, stub_worker,
+                                  redis_state_backend, redis_cancel_backend):
+        stub_broker.add_middleware(Cancel(backend=redis_cancel_backend))
+        stub_broker.add_middleware(MessageState(backend=redis_state_backend))
+        remoulade.declare_actors([TestMessageState.add])
+        stub_worker.pause()
```
Comment (antoinerabany) [inline tests/middleware/test_message_state.py:58]: If you take a look at other tests in the app, you can see that there is some comment explaining what's happening. It's not mandatory but here you can add a comment explaining why you pause the worker (as it's not simple)

File: tests/middleware/test_message_state.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+        msg = TestMessageState.add.send(5, 7)
+        assert redis_state_backend._get(msg.message_id)['state'] == "Pending"
+
+    def test_success_state_message(self, local_broker, redis_state_backend):
+        local_broker.declare_actor(TestMessageState.add)
+        local_broker.add_middleware(MessageState(backend=redis_state_backend))
+        msg = TestMessageState.add.send(5, 7).result
+        assert redis_state_backend._get(msg.message_id)['state'] == "Success"
+
+    def test_started_state_message(self, local_broker, redis_state_backend):
```
Comment (antoinerabany) [inline tests/middleware/test_message_state.py:37]: Do not use a local broker. But a stub broker.
you can test it like this, it's not perfect but enougth:
```
    def test_started_state_message(self, stub_broker, stub_worker, state_middleware):
        
        @remoulade.actor
        def wait():
            time.sleep(0.3)

        stub_broker.declare_actor(wait)

        msg = wait.send()
        time.sleep(0.1)

        assert state_middleware.backend._get(msg.message_id)['state'] == "Started"
```
I can explain it to you if you want

File: tests/middleware/test_message_state.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+        local_broker.add_middleware(MessageState(backend=redis_state_backend))
+        msg = TestMessageState.add.send(5, 7)
+        local_broker.emit_before("process_message", msg)
+        assert redis_state_backend._get(msg.message_id)['state'] == "Started"
+
+    def test_failure_state_message(self, stub_broker, stub_worker, redis_state_backend):
+        stub_broker.declare_actor(TestMessageState.add)
+        stub_broker.add_middleware(MessageState(backend=redis_state_backend))
+        remoulade.declare_actors([TestMessageState.error])
+        msg = TestMessageState.error.send()
```
Comment (antoinerabany) [inline tests/middleware/test_message_state.py:48]: no need to define error as staticmethod as it's only used here.

File: remoulade/state/__init__.py [hunk 5]
```diff
@@ -0,0 +1 @@
+__all__ = ["State", "StateBackend", "MessageState"]
```
Comment (antoinerabany) [inline remoulade/state/__init__.py:1]: you are missing the imports here. when you have a __all__ like this all name in __all__ must be imported.

File: tests/middleware/test_message_state.py (hunk 3 above)
Comment (xdanielsb) [reply tests/middleware/test_message_state.py]: cool, thank u

File: remoulade/state/middleware.py [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+        self.backend = backend
+        self.state_ttl = state_ttl
+
+    def save(self, message, state):
+        key = message.message_id
+        self.backend.store_result(key,
+                                  State(state),
+                                  self.state_ttl)
+
+    def after_enqueue(self, broker, message, delay):
```
[the remark this answers is not in the document]
Comment (xdanielsb) [reply remoulade/state/middleware.py:33]: I think is perfect, in that way we can have a catalogue of default states

File: remoulade/state/backend.py [hunk 7]
```diff
@@ -0,0 +1,51 @@
+from ..encoder import Encoder
+
+
+class State:
```
[the remark this answers is not in the document]
Comment (xdanielsb) [reply remoulade/state/backend.py]: I added the variable arguments, do you think is enough to keep it?

File: remoulade/state/backend.py [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+    Canceled = "Canceled"
+    Failure = "Failure"
+
+    @classmethod
+    def has_value(cls, value):
+        return value in cls._value2member_map_
+
+
+#: A type alias representing states in the database
+class State(namedtuple("State", ("name", "args"))):
```
Comment (antoinerabany) [inline remoulade/state/backend.py:22]: I made some changes to State to make sure we serialize it and deserialize it well. 
```
class State(namedtuple("State", ("name", "args"))):
    """Catalog Class, that storages the state
        Parameters:
            name: Name of the state
            args: List of arguments in the state(name)
    """
    def __new__(cls, name, args):
        if name not in CatalogStates:
            raise InvalidStateError("The {} State is not defined".format(name))
        return super().__new__(cls, name, args)

    @classmethod
    def from_dict(cls, name, *args, **kwargs):
        return cls(name=CatalogStates[name], *args, **kwargs)

    def asdict(self):
        return {
            **self._asdict(),
            'name': self.name.value
        }
```

This way we are you to have a Enum in the state, and we should't need all these .name everywhere.
I'm not sure it should be "name" maybe "state" would be better ? what do you think ?

File: remoulade/state/middleware.py [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+                    in the database, default 2MB
+    """
+
+    def __init__(self, backend, state_ttl=3600, max_size=2e6):
+        self.backend = backend
+        self.state_ttl = state_ttl
+        self.max_size = max_size
+
+    def save(self, message, state):
+        args = message.args
```
Comment (antoinerabany) [inline remoulade/state/middleware.py:24]: you need to add `message.kwargs` too. but nice !

File: remoulade/state/middleware.py [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+        if sys.getsizeof(args) > self.max_size:
+            # Arguments exceed maximum size to display
+            #  do not save them.
+            args = []
+        self.backend.set_state(
+            message.message_id, State(state, args).asdict(), self.state_ttl
+        )
+
+    def after_enqueue(self, broker, message, delay):
+        self.save(message, state=CatalogStates.Pending.name)
```
Comment (antoinerabany) [inline remoulade/state/middleware.py:34]: After the change i made to state e should't have to use all these  `.name`

File: tests/middleware/test_message_state.py [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+
+
+class TestMessageState:
+    """Class to test the middleware
+    MessageState
+    """
+
+    @staticmethod
+    @remoulade.actor
+    def do_work(x):
```
Comment (antoinerabany) [inline tests/middleware/test_message_state.py]: dont need the x argument for do_work.

File: tests/middleware/test_message_state.py [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+        stub_broker.declare_actor(TestMessageState.do_work)
+        msg = TestMessageState.do_work.send(5)
+        stub_broker.join(TestMessageState.do_work.queue_name)
+        stub_worker.join()
+        assert (
+            state_backend.get_state(msg.message_id)["name"] == CatalogStates.Skipped.name
+        )
+
+    @pytest.mark.parametrize(
+        "ttl, state, result_type",
```
Comment (antoinerabany) [inline tests/middleware/test_message_state.py:103]: I think the state parameter is usefull, but you tests are nice !

File: remoulade/state/backends/stub.py [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+            if time.monotonic() > data["expiration"]:
+                self._delete(message_key)
+                data = None
+            else:
+                data = data["state"]
+        return data
+
+    def set_state(self, message_id, state, ttl):
+        message_key = self._build_message_key(message_id)
+        payload = {"state": state, "expiration": ttl}
```
Comment (antoinerabany) [inline remoulade/state/backends/stub.py:26]: you need to store expiration: time.monotonic() + ttl: here and then in get_state check that `time.monotonic() > data["expiration"] `.

File: tests/middleware/test_message_state.py [hunk 14]
```diff
@@ -0,0 +1,134 @@
...
+from remoulade.state.middleware import MessageState
+
+
+class TestMessageState:
+    """Class to test the middleware
+    MessageState
+    """
+
+    def test_pending_state_message(
+        self, stub_broker, state_backend, state_middleware, do_work
```
Comment (antoinerabany) [inline tests/middleware/test_message_state.py:18]: so cool ! If think you don't need to specify `state_backend` and access it via `state_middleware.backend`

File: tests/conftest.py [hunk 15]
```diff
@@ -165,6 +166,48 @@ def redis_result_backend():
...
+    broker.add_middleware(middleware)
+    return middleware
+
+
+@pytest.fixture
+def do_work():
+    broker = remoulade.get_broker()
+
+    @remoulade.actor
+    def do_work(x):
```
Comment (antoinerabany) [inline tests/conftest.py:204]: There should not be an argument for do_work. It should be simplest possible. Use another actor for your last test.

File: tests/middleware/test_message_state.py [hunk 16]
```diff
@@ -0,0 +1,134 @@
...
+        stub_broker.add_middleware(MessageState(backend=state_backend, state_ttl=ttl))
+        stub_broker.declare_actor(wait)
+        msg = wait.send()
+        time.sleep(0.1)
+        data = state_backend.get_state(msg.message_id)
+        # if the ttl is  greater than the expiration, the data should be None
+        assert type(data) == type(result_type)
+
+    @pytest.mark.parametrize(
+        "msize,x,expected", [pytest.param(0, 5, []), pytest.param(1000, 5, [5])],
```
Comment (antoinerabany) [inline tests/middleware/test_message_state.py:125]: use full name when possible: `max_size`, actually x is not changing so no need to parametrize it.

File: remoulade/state/backend.py [hunk 17]
```diff
@@ -0,0 +1,76 @@
+from collections import namedtuple
+from enum import Enum
+
+from ..encoder import Encoder
+from .errors import InvalidStateError
+
+
+class CatalogStates(Enum):
```
[the remark this answers is not in the document]
Comment (xdanielsb) [reply remoulade/state/backend.py:8]: yes, of course

File: remoulade/state/backend.py (hunk 8 above)
Comment (xdanielsb) [reply remoulade/state/backend.py:22]: super cool

File: remoulade/state/backends/stub.py (hunk 13 above)
Comment (xdanielsb) [reply remoulade/state/backends/stub.py:26]: cool perfect

File: remoulade/state/middleware.py [hunk 18]
```diff
@@ -0,0 +1,52 @@
...
+    """
+
+    def __init__(self, backend, state_ttl=3600, max_size=2e6):
+        self.backend = backend
+        self.state_ttl = state_ttl
+        self.max_size = max_size
+
+    def save(self, message, state):
+        args = message.args
+        kwargs = message.kwargs
```
Comment (antoinerabany) [inline remoulade/state/middleware.py:25]: you need to check the size of both args and kwargs:
`sys.getsizeof(args) + sys.getsizeof(kwargs)` and set kwargs to `{}`.

File: remoulade/state/backends/redis.py [hunk 19]
```diff
@@ -0,0 +1,33 @@
...
+        client=None,
+        url=None,
+        **parameters
+    ):
+        super().__init__(namespace=namespace, encoder=encoder)
+        if url:
+            parameters["connection_pool"] = redis.ConnectionPool.from_url(url)
+        self.client = client or redis.Redis(**parameters)
+
+    def get_state(self, message_id):
```
Comment (antoinerabany) [inline remoulade/state/backends/redis.py:21]: I've not been very clear I think. `get_state` should return a state object according to the superclass.  And set state should get a state object. So you will need to call `from_dict` and `to_dict` inside these functions.

Do you about mypy, it's type checking for python: it would have caught this. Maybe we can add a mypy step in the CI in a later PR. Can you create an issue about this ?

https://mypy.readthedocs.io/en/stable/

File: remoulade/state/backend.py [hunk 20]
```diff
@@ -0,0 +1,82 @@
+from collections import namedtuple
+from enum import Enum
+
+from ..encoder import Encoder
+from .errors import InvalidStateError
+
+
+class StatesEnum(Enum):
```
Comment (antoinerabany) [inline remoulade/state/backend.py:8]: Sorry but i think we can do better name for this: as the attribute of the state is name we could name it  `StateNamesEnum` ? would be more clear right ?

File: remoulade/state/backends/redis.py (hunk 19 above)
Comment (xdanielsb) [reply remoulade/state/backends/redis.py:21]: really cool, thanks

File: remoulade/state/backend.py (hunk 20 above)
Comment (xdanielsb) [reply remoulade/state/backend.py:8]: the clear the better, thanks

File: remoulade/state/middleware.py (hunk 18 above)
Comment (xdanielsb) [reply remoulade/state/middleware.py:25]: I modified just a little bit just in case, one of them is not too big to show it:
 if sys.getsizeof(args) then max_size then args = []
 if sys.getsizeof(kargs) then max_size then kwargs = {}

File: docs/source/changelog.rst [hunk 21]
```diff
@@ -310,10 +333,7 @@ Fixed
 
 * pipe_ignore was not recovered from right message
 
-.. _Unreleased: https://github.com/wiremind/remoulade/compare/v0.20.0...HEAD
-.. _0.20.0: https://github.com/wiremind/remoulade/releases/tag/v0.20.0
-.. _0.19.0: https://github.com/wiremind/remoulade/releases/tag/v0.19.0
-.. _0.18.3: https://github.com/wiremind/remoulade/releases/tag/v0.18.3
+.. _Unreleased: https://github.com/wiremind/remoulade/compare/v0.18.1...HEAD
```
Comment (antoinerabany) [inline docs/source/changelog.rst:336]: why did you remove those lines ?

File: remoulade/state/backend.py [hunk 22]
```diff
@@ -0,0 +1,83 @@
...
+            args: List of arguments in the state(name)
+    """
+    def __new__(cls, name, args, kwargs):
+        if not any(state for state in StateNamesEnum if state.name == name):
+            raise InvalidStateError("The {} State is not defined".format(name))
+        return super().__new__(cls, name, args, kwargs)
+
+    @classmethod
+    def from_dict(cls, name, *args, **kwargs):
+        print(name, args, kwargs)
```
Comment (antoinerabany) [inline remoulade/state/backend.py:34]: no print
